src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,125 @@
 package core.basesyntax;
 
+import java.util.Arrays;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size;
+
+    public ArrayList() {
+        elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        checkCapacity();
+        elements[size++] = value;
     }
 
     @Override
     public void add(T value, int index) {
-
+        checkIndexRangeInAdd(index);
+        checkCapacity();
+        checkElementByIndex(index);
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            checkCapacity();
+            elements[size] = list.get(i);
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkIndex(index);
+        elements[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        final Object removedElement = elements[index];
+        Object[] copyElements = new Object[elements.length - 1];
+        System.arraycopy(elements, 0, copyElements, 0, index);
+        System.arraycopy(elements, index + 1, copyElements, index, elements.length - index - 1);
+        elements = copyElements;
```

Comment:
Creating a new array with a reduced size on each removal is inefficient. Instead, you should shift the elements to overwrite the removed element and then nullify the last element if necessary.